delphi-frontend/src/main/java/au/com/integradev/delphi/type/intrinsic/IntrinsicReturnType.java

```diff
@@ -118,4 +142,99 @@ public Type getReturnType(List<Type> arguments) {
       return TypeFactory.unknownType();
     }
   }
+
+  private static final class ConcatReturnType extends IntrinsicReturnType {
+    private final TypeFactory typeFactory;
+
+    public ConcatReturnType(TypeFactory typeFactory) {
+      this.typeFactory = typeFactory;
+    }
+
+    @Override
+    public Type getReturnType(List<Type> arguments) {
+      Type result = null;
+      List<Type> elementTypes = new ArrayList<>();
+
+      for (Type argument : arguments) {
+        if (argument.isDynamicArray() || argument.isString() || argument.isVariant()) {
+          result = widenIfNeeded(argument, result);
+        } else if (argument.isArrayConstructor()) {
+          elementTypes.addAll(((ArrayConstructorType) argument).elementTypes());
+        }
+      }
+
+      if (result == null) {
+        result = typeFactory.arrayConstructor(elementTypes);
+      }
+
+      return result;
+    }
+
+    private Type widenIfNeeded(Type newType, @Nullable Type currentType) {
+      if (newType.isChar()) {
+        newType = typeFactory.getIntrinsic(newType.size() == 1 ? ANSISTRING : UNICODESTRING);
```

Comment:
What's the context of using `newType.size() == 1` here? Looks to me like this is only true for AnsiChar?

Reply:
> this is only true for AnsiChar?

We're just abusing this detail to avoid an unnecessary string comparison.
I don't mind changing it for clarity, though.

Thoughts?

Reply:
I think it would be good to change it for clarity. This could also invisibly be broken at some point in the future, since the code's intention differs from its implementation.